File: app/src/main/java/se/jakob/knarkklocka/data/AlarmDao.java
package se.jakob.knarkklocka.data;

import android.arch.lifecycle.LiveData;
import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Delete;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.OnConflictStrategy;
import android.arch.persistence.room.Query;
import android.arch.persistence.room.Update;

import java.util.List;

//Class to allow interaction of database

@Dao
public interface AlarmDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    long insert(Alarm alarm);

    @Update(onConflict = OnConflictStrategy.REPLACE)
    void updateAlarm(Alarm alarm);

    @Delete
    void deleteAlarm(Alarm alarm);

    @Query("DELETE FROM alarm_table")
    void deleteAll();

    @Query("SELECT * FROM alarm_table ORDER BY end_time")
    LiveData<List<Alarm>> loadAllAlarms();

    @Query("SELECT * FROM alarm_table WHERE id = :id")
    LiveData<Alarm> loadLiveAlarmById(long id);

    @Query("SELECT * FROM alarm_table WHERE id = :id")
    Alarm loadAlarmById(long id);

    @Query("SELECT * FROM alarm_table WHERE alarm_state = :state")
    LiveData<Alarm> loadAlarmsByState(int state);

    @Query("SELECT * FROM alarm_table WHERE alarm_state IN (:states) LIMIT 1")
    LiveData<Alarm> loadSingleAlarmByState(int... states);
}

File: app/src/main/java/se/jakob/knarkklocka/data/AlarmDao.kt
package se.jakob.knarkklocka.data

import androidx.lifecycle.LiveData
import androidx.room.*


//Class to allow interaction of database

@Dao
interface AlarmDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insert(alarm: Alarm): Long

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertAll(alarms: List<Alarm>)

    @Update(onConflict = OnConflictStrategy.REPLACE)
    fun updateAlarm(alarm: Alarm)

    @Delete
    fun deleteAlarm(alarm: Alarm)

    @Query("DELETE FROM alarm_table")
    fun deleteAll()

    @Query("SELECT * FROM alarm_table ORDER BY start_time DESC, end_time DESC")
    fun getAllAlarms(): LiveData<List<Alarm>>

    @Query("SELECT * FROM alarm_table ORDER BY start_time DESC, end_time DESC LIMIT 1")
    fun getMostRecentAlarm(): LiveData<Alarm>

    @Query("SELECT * FROM alarm_table WHERE id = :id")
    fun getLiveAlarm(id: Long): LiveData<Alarm>

    @Query("SELECT * FROM alarm_table WHERE id = :id")
    fun getAlarm(id: Long): Alarm

    @Query("SELECT * FROM alarm_table  WHERE alarm_state IN (:states) ORDER BY start_time DESC, end_time DESC")
    fun getLiveAlarmsByStates(vararg states: Int): LiveData<List<Alarm>>

    @Query("SELECT * FROM alarm_table WHERE alarm_state IN (:states)")
    fun getAlarmsByStates(vararg states: Int): List<Alarm>

    @Query("DELETE FROM alarm_table WHERE alarm_state in (:states)")
    fun deleteAlarmsByStates(vararg states: Int)
}
